src/replace.rs

```diff
@@ -0,0 +1,223 @@
+            );
+
+            let mut new_parts = Vec::with_capacity(self.parts.len() + 2);
+
+            // Previous parts
+            if let Some(ps) = self.parts.get(..index_of_part_to_split) {
+                new_parts.extend_from_slice(&ps);
+            }
+
+            // Keep initial data on left side of part
+            if from > part_to_split.start {
+                new_parts.push(Span {
+                    start: part_to_split.start,
+                    end: from,
+                    data: State::Initial,
+                });
+            }
+
+            // New part
+            new_parts.push(Span {
+                start: from,
+                end: up_to_and_including,
+                data: State::Replaced(data.into()),
+            });
+
+            // Keep initial data on right side of part
+            if up_to_and_including < part_to_split.end {
+                new_parts.push(Span {
+                    start: up_to_and_including + 1,
+                    end: part_to_split.end,
+                    data: State::Initial,
+                });
+            }
+
+            // Following parts
+            if let Some(ps) = self.parts.get(index_of_part_to_split + 1..) {
+                new_parts.extend_from_slice(&ps);
+            }
+
+            new_parts
+        };
+
+        self.parts = new_parts;
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use proptest::prelude::*;
+
+    fn str(i: &[u8]) -> &str {
+        ::std::str::from_utf8(i).unwrap()
+    }
+
+    #[test]
+    fn replace_some_stuff() {
+        let mut d = Data::new(b"foo bar baz");
+        d.replace_range(4, 6, b"lol").unwrap();
+        assert_eq!("foo lol baz", str(&d.to_vec()));
+    }
+
+    #[test]
+    fn replace_a_single_char() {
+        let mut d = Data::new(b"let y = true;");
+        d.replace_range(4, 4, b"mut y").unwrap();
+        assert_eq!("let mut y = true;", str(&d.to_vec()));
+    }
+
+    #[test]
+    fn replace_multiple_lines() {
+        let mut d = Data::new(b"lorem\nipsum\ndolor");
+
+        d.replace_range(6, 10, b"lol").unwrap();
+        assert_eq!("lorem\nlol\ndolor", str(&d.to_vec()));
+
+        d.replace_range(12, 17, b"lol").unwrap();
+        assert_eq!("lorem\nlol\nlol", str(&d.to_vec()));
+    }
+
+    #[test]
+    #[should_panic(expected = "Cannot replace slice of data that was already replaced")]
+    fn replace_overlapping_stuff_errs() {
+        let mut d = Data::new(b"foo bar baz");
+
+        d.replace_range(4, 6, b"lol").unwrap();
+        assert_eq!("foo lol baz", str(&d.to_vec()));
+
+        d.replace_range(4, 6, b"lol").unwrap();
+    }
+
+    #[test]
+    #[should_panic(expected = "original data is only 3 byte long")]
+    fn broken_replacements() {
+        let mut d = Data::new(b"foo");
+        d.replace_range(4, 7, b"lol").unwrap();
+    }
+
+    proptest! {
+        #[test]
+        #[ignore]
```

Comment:
How come these are `#[ignore]`? Do they take too long?

Reply:
They take a few seconds each, yeah. I don't really want to run them on each `cargo test`, but I'll add them to CI